src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -0,0 +1,166 @@
+    """
+
+    def __init__(
+        self,
+        group_id,
+        topic=None,
+        commit_batch_size=100,
+        initial_offset_reset='earliest',
+    ):
+        self.group_id = group_id
+        if not topic:
+            topic = settings.KAFKA_SNUBA_QUERY_SUBSCRIPTIONS
+        self.topic = topic
+        cluster_name = settings.KAFKA_TOPICS[topic]['cluster']
+        self.bootstrap_servers = settings.KAFKA_CLUSTERS[cluster_name]['bootstrap.servers']
+        self.commit_batch_size = commit_batch_size
+        self.initial_offset_reset = initial_offset_reset
+
+    def run(self):
+        logger.debug('Starting snuba query subscriber')
+
+        conf = {
+            'bootstrap.servers': self.bootstrap_servers,
+            'group.id': self.group_id,
+            'session.timeout.ms': 6000,
+            'auto.offset.reset': self.initial_offset_reset,
+            'enable.auto.commit': 'false',
+            'enable.auto.offset.store': 'true',
+            'enable.partition.eof': 'false',
+            'default.topic.config': {'auto.offset.reset': self.initial_offset_reset},
+        }
+        consumer = Consumer(conf)
+        consumer.subscribe([self.topic])
+        try:
+            i = 0
+            while True:
+                message = consumer.poll(0.1)
+                if message is None:
+                    continue
+
+                error = message.error()
+                if error is not None:
+                    raise KafkaException(error)
+
+                i = i + 1
+
+                self.handle_message(message)
+
+                if i % self.commit_batch_size == 0:
+                    logger.debug('Committing offsets')
+                    consumer.commit()
+        except KeyboardInterrupt:
+            pass
+        logger.debug('Committing offsets and closing consumer')
+        consumer.commit()
+        consumer.close()
+
+    def handle_message(self, message):
+        """
+        Parses the value from Kafka, and if valid passes the payload to the callback defined by the
+        subscription. If the subscription has been removed, or no longer has a valid callback then
+        just log metrics/errors and continue.
+        :param message:
+        :return:
+        """
+        try:
+            contents = self.parse_message_value(message.value())
+        except InvalidMessageError:
+            # If the message is in an invalid format, just log the error
+            # and continue
+            logger.exception('Subscription update could not be parsed')
+            return
+
+        try:
+            subscription = QuerySubscription.objects.get(
+                subscription_id=contents['subscription_id'])
+        except QuerySubscription.DoesNotExist:
+            metrics.incr('snuba_query_subscriber.subscription_doesnt_exist')
+            logger.error('Received subscription update, but subscription does not exist')
```

Comment:
Is this something we want alerts on long term? If a subscription is killed while a consumer is behind what do we need to do?

Reply:
We might, just to catch anything anomalous. If a subscription is ever killed I think it's safe to assume that the user doesn't ever want to receive alerts from it, even if those alerts would have occurred before the subscription was deleted.